src/workload.rs

```diff
@@ -601,17 +741,25 @@ impl WorkloadStore {
         // include them in load balancing we do to Services.
         if status == HealthStatus::Healthy {
             for (vip, pl) in &w.virtual_ips {
-                let ip = vip.parse::<IpAddr>()?;
-                for port in &pl.ports {
-                    let service_sock_addr = SocketAddr::from((ip, port.service_port as u16));
-                    self.vips
-                        .entry(service_sock_addr)
-                        .or_default()
-                        .insert((wip, port.target_port as u16));
-                    self.workload_to_vip
-                        .entry(wip)
-                        .or_default()
-                        .insert((service_sock_addr, port.target_port as u16));
+                let vip = vip.parse::<IpAddr>()?;
+
+                let ep = Endpoint {
+                    address: wip,
+                    port: pl.into(),
+                };
+                if let Some(svc) = self.vips.get_mut(&vip) {
+                    svc.endpoints.insert(ep.address, ep);
+                    self.workload_to_vip.entry(wip).or_default().insert(vip);
+                } else {
+                    // Can happen due to ordering issues
+                    trace!("pod has VIP {vip}, but VIP not found");
+                    if let Some(ep_map) = self.staged_vips.get_mut(&vip) {
+                        ep_map.insert(wip, ep);
+                    } else {
+                        let mut ep_map = HashMap::new();
+                        ep_map.insert(wip, ep);
+                        self.staged_vips.insert(vip, ep_map);
```

Comment:
I don't see anywhere where staged_vips is removed when we remove a workload. Can you add a testcase for this?

Reply:
https://github.com/istio/ztunnel/pull/486/commits/a08e65d78de4d32530baba79c0608c404b6073d4 fixes the bug for removing endpoints and tests it